matlab_simulation/05-estimation/kalman_filter/kalman_filter.m

```diff
@@ -0,0 +1,59 @@
+function [mu,S,mup,Sp,K] = kalman_filter(ssm,mu,S,u,y,example,t)
+
+% Performs one iteration of Kalman Filtering
+
+% Inputs:
+%       ssm                 : State space model structure
+%       mu                  : Mean of current state
+%       S                   : Covariance of current state
+%       u                   : Control Input of current time step
+%       y                   : Measurment of current time step
+%       example             : Example under consideration
+%       t                   : Current time step
+% Outputs:
+%       mu                  : Final estimated mean after iteration
+%       S                   : Final estimated covariance after iteration
+%       mup                 : Predicted mean after motion update
+%       Sp                  : Predicted covariance after motion update
+%       K                   : Kalman gain for current time step 
+
+% Unwrap state space model
+A = ssm.A; 
+B = ssm.B;
+C = ssm.C; 
+D = ssm.D;
+R = ssm.R;
+Q = ssm.Q;
+n = ssm.n;
+m = ssm.m;
+
+% Prediction Step: ------------------------------------------------
+% Predict new mean based on motion
+mup = A*mu + B*u;
+% Predict new covariance based on motion
+Sp = A*S*A' + R;
+
+% Measurement Step: ------------------------------------------------
+if example == 3 % exmaple 3 is a multirate kalman filter
+    Qp = Q.Qp;
+    Qv = Q.Qv;
+    Cp = C.Cp;
+    Cv = C.Cv;
+    if (mod(t,10) == 0)
```

Comment:
I would add a comment explaining the if/else workings of how each tenth step has the reduced variance. Possibly make the 10 into a variable in case you want to modify the multi-rate behaviour.